### PR TITLE
CNV 778 - Booting VMs in EFI Mode

### PR DESCRIPTION
**Pull request by mgarrell777**
This PR address JIRA story [CNV-778](https://issues.redhat.com/browse/CNV-778) (https://issues.redhat.com/browse/CNV-778).

Added information to Advanced VM Management about Booting a VM in EFI Mode.

Preview link - https://deploy-preview-28761--osdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/advanced_vm_management/virt-efi-mode-for-vms.html

### CODE REVIEW

File: modules/virt-booting-efi-mode.adoc [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+[id="virt-booting-vms-efi-mode_{context}"]
+= Booting virtual machines in EFI mode
+
+Booting a virtual machine in EFI mode allows for a quicker boot process.
+
+.Procedure
+To boot a VM in EFI Mode:
+
+* Create a .yaml file that defines a VM object or a Virtual Machine Instance (`VMI`) object. Use the firmware section of the example .yaml file:
```
Comment (stu-gott) [inline modules/virt-booting-efi-mode.adoc:13]: Do we refer to these as "yaml files" in general? Not saying this is the wrong term to use. Just wondering if we generally refer directly to the yaml itself.

Also "section" -> "stanza"?

Comment (stu-gott) [reply modules/virt-booting-efi-mode.adoc:13]: Maybe something like:

"Define a VM or Virtual Machine Instance (`VMI`) object with SecureBoot enabled in the firmware stanza"?

**Review by stu-gott — COMMENTED** [review]
One nitpick. Otherwise this looks good!

File: modules/virt-booting-efi-mode.adoc [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  spec:
+    domain:
+      devices:
+        disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      features:
+        acpi: {}
+        smm:
```
Comment (vsibirsk) [inline modules/virt-booting-efi-mode.adoc:33]: I think it should be explicitly mentioned that SMM (System Management Mode) feature needs to be enabled.
Otherwise VM will fail on creation

**Review by vsibirsk — COMMENTED** [review]
In general looks good. Only one small comment

**Comment by mgarrellRH** [comment]
@stu-gott @vsibirsk I have made updates based on your initial feedback. Please take a look and let me know if there's anything else I need to change. Thanks very much.

https://deploy-preview-28761--osdocs.netlify.app/openshift-enterprise/latest/virt/virtual_machines/advanced_vm_management/virt-efi-mode-for-vms.html

**Review by stu-gott — COMMENTED** [review]
LGTM

**Comment by vsibirsk** [comment]
lgtm

File: _topic_map.yml [hunk 3]
```diff
@@ -2349,6 +2349,8 @@ Topics:
 #Advanced virtual machine configuration
     - Name: Automating management tasks
       File: virt-automating-management-tasks
+    - Name: EFI Mode for Virtual Machines
```
Comment (ousleyp) [inline _topic_map.yml:2352]: The title/TOC entry should be in sentence case.

File: modules/virt-about-efi-mode-for-vms.adoc [hunk 4]
```diff
@@ -0,0 +1,17 @@
+// Module included in the following assemblies:
+//
+// * virt-efi-mode-for-vms.adoc
```
Comment (ousleyp) [inline modules/virt-about-efi-mode-for-vms.adoc:3]: This needs to be the full path, starting with the `virt` directory

File: virt/virtual_machines/advanced_vm_management/virt-efi-mode-for-vms.adoc [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[id="virt-efi-mode-for-vms"]
+= EFI mode for virutal machines
```
Comment (ousleyp) [inline virt/virtual_machines/advanced_vm_management/virt-efi-mode-for-vms.adoc:2]: typo: virtual

Comment (ousleyp) [inline virt/virtual_machines/advanced_vm_management/virt-efi-mode-for-vms.adoc:2]: I would suggest putting a gerund in this title

File: virt/virtual_machines/advanced_vm_management/virt-efi-mode-for-vms.adoc [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+include::modules/virt-document-attributes.adoc[]
+include::modules/common-attributes.adoc[]
+:context: virt-efi-mode-for-vms
+toc::[]
+
+A virtual machine (`VM`) can be booted in `EFI` (Extensible Firmware Interface) mode.
```
Comment (ousleyp) [inline virt/virtual_machines/advanced_vm_management/virt-efi-mode-for-vms.adoc:8]: Passive voice + I do not think the backticks are necessary here. Suggestion: "You can boot a virtual machine (VM) in Extensible Firmware Interface (EFI) mode."

File: modules/virt-about-efi-mode-for-vms.adoc [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+[id="virt-about-efi-mode-for-vms_{context}"]
+= About EFI mode for virtual machines
+
+`EFI` (Extensible Firmware Interface) is a firmware interface that initializes the hardware components and starts the operating system stored on a hard disk drive when a computer starts.
```
Comment (ousleyp) [inline modules/virt-about-efi-mode-for-vms.adoc:8]: I would suggest rewriting this sentence for more clarity and user focus, because I don't quite understand what you are trying to say. It might also be better to split it into two sentences. 

Specifically, I'm confused about what "the operating system" means. Is it an OS that is stored on a hard drive versus a different OS? 

s/stored on/that is stored on

File: modules/virt-about-efi-mode-for-vms.adoc [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+It stores all the information about initialization and startup in a .efi file, which is stored on a special partition called EFI System Partition (ESP). The ESP also contains the boot loader programs for the operating system installed on the computer.
```
Comment (ousleyp) [inline modules/virt-about-efi-mode-for-vms.adoc:10]: s/.efi/`.efi`

s/operating system installed/operating system that is installed

File: modules/virt-about-efi-mode-for-vms.adoc [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+The ESP allows `EFI` to directly boot the operating system and save the BIOS self-test process, allowing the computer to boot faster.
```
Comment (ousleyp) [inline modules/virt-about-efi-mode-for-vms.adoc:12]: no backticks around EFI

What does "save" mean here?

File: modules/virt-about-efi-mode-for-vms.adoc [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+[NOTE]
+====
+{VirtProductName} only supports a virtual machine (VM) with Secure Boot when using EFI mode. If Secure Boot is not enabled, the VM crashes repeatedly. However, the VM may not support Secure Boot. Prior to booting a VM, verify the VM’s settings support Secure Boot.
```
Comment (ousleyp) [inline modules/virt-about-efi-mode-for-vms.adoc:16]: Some clarity/translation nits:

s/However, the VM may not/However, some VMs do not

(side note that the IBM style guide says to use "might," not "may")

s/Prior to booting/Before you boot

s/verify the VM's settings support Secure Boot./verify that it supports Secure Boot by checking the VM settings.

File: modules/virt-booting-efi-mode.adoc [hunk 11]
```diff
@@ -0,0 +1,52 @@
+// Module included in the following assemblies:
+//
+// * virt-efi-mode-for-vms.adoc
```
Comment (ousleyp) [inline modules/virt-booting-efi-mode.adoc:3]: full path, starting with `virt` directory

File: modules/virt-about-efi-mode-for-vms.adoc (hunk 9 above)
Comment (ousleyp) [reply modules/virt-about-efi-mode-for-vms.adoc:12]: If faster booting is the main reason for using EFI mode, I think you should mention that earlier in the module.

File: modules/virt-about-efi-mode-for-vms.adoc (hunk 7 above)
Comment (ousleyp) [reply modules/virt-about-efi-mode-for-vms.adoc:8]: I would suggest rewriting the module so that you start by explaining why you might want to use EFI (faster boot), and *then* include the more technical information.

File: modules/virt-booting-efi-mode.adoc [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+[id="virt-booting-vms-efi-mode_{context}"]
+= Booting virtual machines in EFI mode
+
+Booting a virtual machine in EFI mode allows for a quicker boot process.
```
Comment (ousleyp) [inline modules/virt-booting-efi-mode.adoc:8]: s/quicker/faster

(consider including a comparison to a slower boot mode; what is it faster than?)

File: modules/virt-booting-efi-mode.adoc [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+[NOTE]
```
Comment (ousleyp) [inline modules/virt-booting-efi-mode.adoc:10]: If you enable this at the same time as you enable secureBoot, I would note that as a callout in the YAML instead of this note here.

File: modules/virt-booting-efi-mode.adoc [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+====
+{VirtProductName} requires System Management Mode (`SMM`) to be enabled for Secure Boot in EFI mode to occur.
+====
+
+.Procedure
+To boot a VM in EFI Mode:
```
Comment (ousleyp) [inline modules/virt-booting-efi-mode.adoc:16]: This doesn't need an intro

File: modules/virt-booting-efi-mode.adoc [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+. Create a YAML file that defines a VM object or a Virtual Machine Instance (`VMI`) object. Use the firmware stanza of the example YAML file:
++
+.Booting in `EFI` mode with secure boot active
```
Comment (ousleyp) [inline modules/virt-booting-efi-mode.adoc:20]: no backticks

File: modules/virt-booting-efi-mode.adoc [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    domain:
+      devices:
+        disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      features:
+        acpi: {}
+        smm:
+          enabled: true
```
Comment (ousleyp) [inline modules/virt-booting-efi-mode.adoc:39]: I would add callouts on this line and the `secureBoot: true` line, to point out that you should edit these.

File: modules/virt-booting-efi-mode.adoc [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+      firmware:
+        bootloader:
+          efi:
+            secureBoot: true
+...
+----
+
+. Load the YAML file in the Command Line Interface (`CLI`) using the corresponding user operation:
```
Comment (ousleyp) [inline modules/virt-booting-efi-mode.adoc:47]: I would add a prereq to this module:

* Install the OpenShift CLI (`oc`).

Comment (ousleyp) [inline modules/virt-booting-efi-mode.adoc:47]: This is a bit wordy, IMO. I'd change this to something like "Apply the manifest to your cluster by running the following command:"

File: modules/virt-booting-efi-mode.adoc [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
++
+[source,terminal]
+----
+$ oc  create -f my-vm-def-file.yaml
```
Comment (ousleyp) [inline modules/virt-booting-efi-mode.adoc:51]: extra space between `oc` and `create`

(I think we usually do `oc apply`. If it must be `create`, I'd change the previous step to "Create the VM or VMI by running the following command:")

**Review by ousleyp — COMMENTED** [review]
A handful of nits and suggestions, but you're on the right track. Thanks, Matthew!

**Comment by ousleyp** [comment]
Looking good. 👍  
Merging away!
